### PR TITLE
V1 Rewrite

### LINKED ISSUE #65: Handling errors that are thrown in onNoMatch handler
It's a Feature request.

`options.onError` function could catch errors that are thrown in `options.onNoMatch` handler. So in a project we could explicitly throw custom error in onNoMatch function.
As a result, all errors will be processed in а single place.

```
import { NotFoundError } from 'errors/not-found-error';

export const onNoMatch = (): void => {
  throw new NotFoundError();
};

```

### LINKED ISSUE #126: 1.0 Release
This library has been sitting in v0 for awhile. After a period of using it in several production apps, it does not seem to yield any problems. After all, most crux of this library depends on the awesome https://github.com/lukeed/trouter library, so nothing much can really go wrong.

Before the 1.0 release though:
- I want to make sure the APIs are reasonable so that we can avoid breaking changes to it.
- Update documentation to be more general. Even though this is originally built of Next.js, I have been using it on top on `http` very fine. Therefore, it should be about more than Next.js.
- Decide on whether to add several features like routing caching to improve perf. I have been debating this since one of the goal of this library is to stay lightweight.
- Decide on whether to add several utils like middleware concurrent execution

### LINKED ISSUE #146: Support async middleware chains
Hello! I'm trying to implement a response time calculator in a Next API Route package I'm developing, and running into a roadblock.

The middleware stack currently runs synchronously, which prevents middleware from executing code after an async handler has been resolved. Instead, the stack unravels after an `await`ed call, executing code after `next()` in a middleware while doing so. The example below illustrates this.

Would there be any interest in a PR which changes this behavior and adds accompanying documentation?

The simplest method would be to change [`next()`](https://github.com/hoangvvo/next-connect/blob/master/src/index.js#L70) to return the result of it's expression. This would let users form async middleware chains in their applications, resolving the issue described above. It also has the benefit of being a non-breaking change. 

The only major limitation is that async middleware would be all or nothing. Any middleware that fails to `await` or `return` the result of `next()` would break the async chain. This could cause some major confusion, and would need to be well documented. The same behavior can be found in other libraries such as [koa.js](https://koajs.com/#cascading) and the long awaited Express v5, however, so it's not an entirely new concept. 

## Example

```js
import nc from 'next-connect';
import { someAsyncWork } from './another-module';

const middleware = (req, res, next) => {
  console.log('1');

  await next(); // await has no bearing in the current version, however with the proposed solution, it would.

  console.log('4');
};

const handler = async (req, res) => {
  console.log('2');

  const someData = await someAsyncWork();
  res.status(200).json(someData);

  console.log('3');
};

export default nc()
  .use(middleware)
  .get('/', handler);
```
#### Output

| Expected | Actual |
|:-:|:-:|
| 1 | 1 |
| 2 | 2 |
| 3 | **_4_** |
| 4 | **_3_** |

### LINKED ISSUE #170: Running a handler at the end of chain?
does anyone know if it's possible to have a handler that always runs at the end of the chain? A use-case may be a logger middleware that logs to stdout once all middleware have run.  For example if you've got:

```
function onError(err, req, res, next) {
  ...
}

function logger(req,res) {
    // a Pino logger
}

const baseHandler = nc({ onError })
	.use(auth)
   .use(somethingElse);

// Some API using base handler
baseHandler
  .get("/api/foo", (req, res, next) => {
		// do something here
  });

```

How do you add `logger` so that it logs every req/req as part of the middleware chain but ran last? I don't want to make every `get/post/put...` responsible for its own logging. How are you peeps doing it?

### LINKED ISSUE #184: Handle error in getServerSideProps
I want to reuse same error handler from APIs in getServerSideProps. I looked carefully in Readme doc and I see that neither `handler.use(base, ...fn)` can receive error middleware, nor `handler.run(req, res)` triggers main `onError` handler. How can I reuse same API error handler for getServerSideProps? 

I think it would be more elegant to either can have `use(onErrorMiddleware)` or an option to trigger the default API `onError` handler.

Here is what I have:

```ts
// main API handler
export const ncOptions = {
  onError(error: Error, req: NextApiRequest, res: NextApiResponse) {
    handleError(error, req, res);
  },
  onNoMatch(req: NextApiRequest, res: NextApiResponse) {
    const error = new ApiError(`Method '${req.method}' not allowed`, 405);
    //not in api, handle it manually
    handleError(error, req, res);
  },
};

export type NextApiRequestWithResult = NextApiRequest & { result: any };

type NextReq = IncomingMessage & {
  cookies: NextApiRequestCookies;
};

// SSR handler that receives callback and returns result
export const ssrNcHandler = async (
  req: NextReq,
  res: ServerResponse,
  callback: () => Promise<any>
) => {
  const base = () => {
    const handler = nc<NextApiRequestWithResult, NextApiResponse>(ncOptions).use(
      async (req, res, next) => {
        req.result = await callback();
        next();
      } // cant I pass here onErrorMiddleware?
    );

    return handler;
  };

  const _req = req as NextApiRequestWithResult;
  const _res = res as NextApiResponse;

  try {
    await base().run(_req, _res);

    return _req.result;
  } catch (error) {
    handleError(error, _req, _res); // handle error here, log etc...
    return null;
  }
};
```

Usage:

```ts
export const getServerSideProps: GetServerSideProps = async ({ req, res, params }) => {
  const id = Number(params?.id);

  const callback = async () => await getPostWithAuthorById(id);
  const post = (await ssrNcHandler(req, res, callback)) as PostWithUser;

  if (!post) {
    return {
      notFound: true,
    };
  }
...
```

### LINKED ISSUE #188: how to order the response [onNoMatch]
when i send request to **post** method with **get** method it works and i don't even use get method on this url

my code 

how can i get 404 page not found on  [onNoMatch] instead of this response
```
{
  "validationErrors": {
      "email": "This Field Is Required",
      "password": "This Field Is Required"
    }
}
```

### LINKED ISSUE #189: Cannot access '__WEBPACK_DEFAULT_EXPORT__' before initialization
Hey team! I'm working through adding `next-connect` to all of my API pages in a Next.js project. I'm using the factory pattern that I've seen outlined in other issues:

```ts
import nc from 'next-connect';

export default function nextConnectHandler() {
  return nc<NextApiRequest, NextApiResponse>({
    onError(error: Error, req: NextApiRequest, res: NextApiResponse) {
      console.log({ error });
      res.status(501).json({ error: `${error.message}` });
    },
    onNoMatch(req: NextApiRequest, res: NextApiResponse) {
      res.status(405).json({ error: `Method '${req.method}' Not Allowed` });
    },
  });
}
```

Then, from an API route, I can call it like this:

```ts
// /api/foo.ts
import nextConnectHandler from 'utils/nextConnectHandler';

const apiRoute = nextConnectHandler();

apiRoute.get(async (req, res) => {
  // some logic
});

export default apiRoute;
```

On the client, results are sporadic. Sometimes my API requests resolve just fine, but other times I'm getting waves of this error:

Screen Shot 2022-05-16 at 22 37 20@2x

```
error - (api)/utils/nextConnectHandler.ts (23:44) @ nextConnectHandler
TypeError: (0 , next_connect__WEBPACK_IMPORTED_MODULE_1__.default) is not a function
```

Have you encountered something like this before? Googling and issue search on this repo didn't turn up anything quite like it.

I have also tried:

- `rm -rf node_modules yarn.lock && yarn` to make sure I'm using latest package
- delete `.next` directory and restart the app with `vercel dev` to recompile

Interestingly, I can get some of my pages to load and fetch data (correctly!) when server rendered. But as soon as the page renders and the React hydrates, I have some `swr` calls to my API that are failing with the above error.

Here are my versions:
```json
"next-connect": "^0.12.2",
"next": "^12.1.6",
```

### LINKED ISSUE #194: redirecting in getServerSideProps middleware
Hello,

the documentation indicates that we can't redirect using res.redirect() in a middleware.

```
DO NOT use response function like res.(s)end or res.redirect inside getServerSideProps.
// page/index.js
const handler = nc()
  .use((req, res) => {
    // BAD: res.redirect is not a function (not defined in `getServerSideProps`
    res.redirect("foo");
  })
  .use((req, res) => {
    // BAD: `getServerSideProps` gives undefined behavior if we try to send response
    res.end("bar");
  });

export async function getServerSideProps({ req, res }) {
  await handler.run(req, res);
  return {
    props: {},
  };
}
```

I am making an authentication middleware that should redirect to the login page when the token is expired and stop the middleware chain if the user is not authenticated. But I can't seem to find a solution to handle that from the middlewares.
How should I do this ?

Thanks in advance !

### PR DESCRIPTION
**Pull request by hoangvvo**
This is the rewrite of next-connect. There are many breaking changes. See [README.md](https://github.com/hoangvvo/next-connect/tree/v1-rewrite) for more detail.

The motivation of this rewrite is to:
- provide a more stable/straight-forward implementation. avoid [this hacky solution](https://github.com/hoangvvo/next-connect/blob/v0/src/index.js#L15-L26). Now the handler will resolve based on pure promise resolution.
- new async paradigm like Koa for more flexibility. Close #146, Close #170, Close #184.
- correct some behaviors. 
  - in handler(), if there is no matching .method() (even if there are matchign .use()`, nothing should be executed. close #188. 
  - onError should also be triggered when throwing in onNoMatch. close #65
- [forward returned value from handlers](https://github.com/hoangvvo/next-connect/blob/v1-rewrite/test/node.test.ts#L48), which enables some interesting use cases. close #194
- Move to named export to avoid some issues with bundler. Close #189. We also have full support for ESM/CJS
- Supports [Edge Runtime](https://edge-runtime.vercel.sh), [Edge API Routes](https://nextjs.org/docs/api-routes/edge-api-routes), [Middleware](https://nextjs.org/docs/advanced-features/middleware). The [agnostic router](https://github.com/hoangvvo/next-connect/blob/v1-rewrite/src/router.ts) allows us to implement this easily. (will be done in a separate pr)

With this, we will be heading for 1.0! [0.x version will still be supported](https://github.com/hoangvvo/next-connect/tree/v0) (I am planning to make some semi-breaking changes to correct some behaviors, esp. the two in the last bullet point above, as well as fixing some bugs) Close #126

Check out the example: https://github.com/hoangvvo/next-connect/tree/v1-rewrite/examples/nextjs

### CODE REVIEW

**Comment by merlindru** [comment]
What's required to get this landed?

Could we get an npm release? `npm i next-connect@next` or `@beta` would be cool!

File: src/router.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    }
+    return this;
+  }
+  public all: RouteShortcutMethod<this, H> = this.add.bind(this, "");
+  public get: RouteShortcutMethod<this, H> = this.add.bind(this, "GET");
+  public head: RouteShortcutMethod<this, H> = this.add.bind(this, "HEAD");
+  public post: RouteShortcutMethod<this, H> = this.add.bind(this, "POST");
+  public put: RouteShortcutMethod<this, H> = this.add.bind(this, "PUT");
+  public patch: RouteShortcutMethod<this, H> = this.add.bind(this, "PATCH");
+  public delete: RouteShortcutMethod<this, H> = this.add.bind(this, "DELETE");
```
Comment (merlindru) [inline src/router.ts:60]: Why is this using `this.add.bind(this)` over `() => this.add(this)` syntax? Just preference?

**Comment by hoangvvo** [comment]
> Awesome!!! Thank you for your hard work! slightly_smiling_face
> 
> What's required to get this landed?
> 
> Could we get an npm release? `npm i next-connect@next` or `@beta` would be cool!

This should be landing soon :rocket:

**Comment by hoangvvo** [comment]
Published as `next`!

```
npm i next-connect@next
```

**Pull request merged**